{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extra Credit\n",
    "## Project option 1 - Drug annotation of 23andme report\n",
    "\n",
    "\n",
    "Columns in the **23andme_v5_hg19_ref.txt** file are:\n",
    "- CHR: the chromosome number\n",
    "- POS: the position of the variant - position where the difference in nucleotides is found\n",
    "- dbSNP_ID: the variant identifier in the dbSNP database (https://www.ncbi.nlm.nih.gov/snp/) (this field will contain a dot . when the SNP is not available in the dbSNP database)\n",
    "- **ALLELE_23andme** : ALLELE: the nucleotides found at that position on the pair of chromosomes\n",
    "\n",
    "\n",
    "Columns in the **var_drug_ann.tsv** file are:\n",
    "- Variant Annotation ID: unique ID number for each variant/drug annotation    \n",
    "- **dbSNP ID** : Variant/Haplotypes: dbSNP ID or haplotype(s)\n",
    "- **GENE_SYMBOL** : Gene: HGNC symbol\n",
    "- **DRUG_NAME** : Drug(s): Drug name\n",
    "- **PMID** : PubMed identifier\n",
    "- **PHENOTYPE_CATEGOERY**: Phenotype Category: options [efficacy, toxicity, dosage, metabolism/PK other]\n",
    "- **SIGNIFICANCE**: Significance: yes or no – determined by if the author stated the association was significant\n",
    "- **NOTES** :Notes: curator notes field\n",
    "- **SENTENCE**: Sentence: structured sentence\n",
    "- **ALLELE_PharmGKB**: Alleles: variant alleles in annotation\n",
    "- Specialty Population: tags for any special populations this annotation is relevant to (e.g. pediatric)\n",
    "\n",
    "And other columns: Metabolizer types, isPlural, Is/Is Not associated, Direction of effect, PD/PK terms, Multiple drugs And/or, Population types, Population Phenotypes or diseases, Multiple phenotypes or diseases And/or, Comparison Allele(s) or Genotype(s), Comparison Metabolizer types"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Q1. Map/merge the 23andme file and the variant-drug annotation file based on dbSNP_ID (also known as rsID).\n",
    "\n",
    "The merged result should have the following columns:\n",
    "\n",
    "dbSNP_ID, GENE_SYMBOL, DRUG_NAME, PMID, PHENOTYPE_CATEGORY, SIGNIFICANCE, NOTES, SENTENCE, ALLELE_PharmGKB (variant alleles in annotation), ALLELE_23andme (variant alleles in 23andme file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>dbSNP_ID</th>\n",
       "      <th>ALLELE_23andme</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69869</td>\n",
       "      <td>rs548049170</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>74792</td>\n",
       "      <td>rs13328684</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHR    POS     dbSNP_ID ALLELE_23andme\n",
       "0  chr1  69869  rs548049170              T\n",
       "1  chr1  74792   rs13328684              G"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# loading the 23andme data into a dataframe\n",
    "twentythree_df = pd.read_csv('23andme_v5_hg19_ref.txt', sep='\\t', comment='#', header=None, names=['CHR', 'POS', 'dbSNP_ID', 'ALLELE_23andme']).dropna()\n",
    "twentythree_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant Annotation ID</th>\n",
       "      <th>Variant/Haplotypes</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Drug(s)</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Phenotype Category</th>\n",
       "      <th>Significance</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Sentence</th>\n",
       "      <th>Alleles</th>\n",
       "      <th>...</th>\n",
       "      <th>isPlural</th>\n",
       "      <th>Is/Is Not associated</th>\n",
       "      <th>Direction of effect</th>\n",
       "      <th>PD/PK terms</th>\n",
       "      <th>Multiple drugs And/or</th>\n",
       "      <th>Population types</th>\n",
       "      <th>Population Phenotypes or diseases</th>\n",
       "      <th>Multiple phenotypes or diseases And/or</th>\n",
       "      <th>Comparison Allele(s) or Genotype(s)</th>\n",
       "      <th>Comparison Metabolizer types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1451834452</td>\n",
       "      <td>CYP3A4*1, CYP3A4*17</td>\n",
       "      <td>CYP3A4</td>\n",
       "      <td>nifedipine</td>\n",
       "      <td>15634941</td>\n",
       "      <td>Other, Metabolism/PK</td>\n",
       "      <td>not stated</td>\n",
       "      <td>in vitro expression of the recombinant CYP3A4*...</td>\n",
       "      <td>CYP3A4 *17 is associated with decreased metabo...</td>\n",
       "      <td>*17</td>\n",
       "      <td>...</td>\n",
       "      <td>Is</td>\n",
       "      <td>Associated with</td>\n",
       "      <td>decreased</td>\n",
       "      <td>metabolism of</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>*1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1451159680</td>\n",
       "      <td>rs5031016</td>\n",
       "      <td>CYP2A6</td>\n",
       "      <td>warfarin</td>\n",
       "      <td>22248286</td>\n",
       "      <td>Dosage</td>\n",
       "      <td>no</td>\n",
       "      <td>No association was found between this variant ...</td>\n",
       "      <td>Allele G is not associated with increased dose...</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>Is</td>\n",
       "      <td>Not associated with</td>\n",
       "      <td>increased</td>\n",
       "      <td>dose of</td>\n",
       "      <td>NaN</td>\n",
       "      <td>in people with</td>\n",
       "      <td>Other:an international normalized ratio (INR) ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Variant Annotation ID   Variant/Haplotypes    Gene     Drug(s)      PMID  \\\n",
       "0             1451834452  CYP3A4*1, CYP3A4*17  CYP3A4  nifedipine  15634941   \n",
       "1             1451159680            rs5031016  CYP2A6    warfarin  22248286   \n",
       "\n",
       "     Phenotype Category Significance  \\\n",
       "0  Other, Metabolism/PK   not stated   \n",
       "1                Dosage           no   \n",
       "\n",
       "                                               Notes  \\\n",
       "0  in vitro expression of the recombinant CYP3A4*...   \n",
       "1  No association was found between this variant ...   \n",
       "\n",
       "                                            Sentence Alleles  ... isPlural  \\\n",
       "0  CYP3A4 *17 is associated with decreased metabo...     *17  ...       Is   \n",
       "1  Allele G is not associated with increased dose...       G  ...       Is   \n",
       "\n",
       "  Is/Is Not associated Direction of effect    PD/PK terms  \\\n",
       "0      Associated with           decreased  metabolism of   \n",
       "1  Not associated with           increased        dose of   \n",
       "\n",
       "  Multiple drugs And/or Population types  \\\n",
       "0                   NaN              NaN   \n",
       "1                   NaN   in people with   \n",
       "\n",
       "                   Population Phenotypes or diseases  \\\n",
       "0                                                NaN   \n",
       "1  Other:an international normalized ratio (INR) ...   \n",
       "\n",
       "  Multiple phenotypes or diseases And/or Comparison Allele(s) or Genotype(s)  \\\n",
       "0                                    NaN                                  *1   \n",
       "1                                    NaN                                   A   \n",
       "\n",
       "  Comparison Metabolizer types  \n",
       "0                          NaN  \n",
       "1                          NaN  \n",
       "\n",
       "[2 rows x 22 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# loading the var_drug data into a dataframe\n",
    "# rouge line was taken out in vim manually\n",
    "drug_df = pd.read_csv('var_drug_ann.tsv', sep='\\t', comment='#')\n",
    "drug_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbSNP_ID</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>PMID</th>\n",
       "      <th>PHENOTYPE_CATEGORY</th>\n",
       "      <th>SIGNIFICANCE</th>\n",
       "      <th>NOTES</th>\n",
       "      <th>SENTENCE</th>\n",
       "      <th>ALLELE_PharmGKB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CYP3A4*1, CYP3A4*17</td>\n",
       "      <td>CYP3A4</td>\n",
       "      <td>nifedipine</td>\n",
       "      <td>15634941</td>\n",
       "      <td>Other, Metabolism/PK</td>\n",
       "      <td>not stated</td>\n",
       "      <td>in vitro expression of the recombinant CYP3A4*...</td>\n",
       "      <td>CYP3A4 *17 is associated with decreased metabo...</td>\n",
       "      <td>*17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs5031016</td>\n",
       "      <td>CYP2A6</td>\n",
       "      <td>warfarin</td>\n",
       "      <td>22248286</td>\n",
       "      <td>Dosage</td>\n",
       "      <td>no</td>\n",
       "      <td>No association was found between this variant ...</td>\n",
       "      <td>Allele G is not associated with increased dose...</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CYP2D6 ultrarapid metabolizer genotype</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>tramadol</td>\n",
       "      <td>18204346</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>Median (+)R,R-tramadol area under the curve wa...</td>\n",
       "      <td>CYP2D6 ultra-metabolizer genotype is associate...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CYP2C9*11</td>\n",
       "      <td>CYP2C9</td>\n",
       "      <td>warfarin</td>\n",
       "      <td>33350885</td>\n",
       "      <td>Dosage</td>\n",
       "      <td>not stated</td>\n",
       "      <td>\"This case suggests that CYP2C9 *11/*11 carrie...</td>\n",
       "      <td>CYP2C9 *11/*11 is associated with decreased do...</td>\n",
       "      <td>*11/*11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CYP2B6*1, CYP2B6*18</td>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>efavirenz</td>\n",
       "      <td>16495778</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>Please note that in the paper the allele was r...</td>\n",
       "      <td>CYP2B6 *1/*18 is associated with increased con...</td>\n",
       "      <td>*1/*18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11597</th>\n",
       "      <td>rs2242480</td>\n",
       "      <td>CYP3A4</td>\n",
       "      <td>tacrolimus</td>\n",
       "      <td>38004558</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>\"CYP3A4*18B genotypes were also significant co...</td>\n",
       "      <td>Genotypes CT + TT is associated with increased...</td>\n",
       "      <td>CT + TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11598</th>\n",
       "      <td>rs7590387</td>\n",
       "      <td>RAMP1</td>\n",
       "      <td>erenumab</td>\n",
       "      <td>34965002</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>no</td>\n",
       "      <td>\"RAMP1 rs7590387 was found to confer a lower p...</td>\n",
       "      <td>Allele G is associated with decreased clinical...</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11599</th>\n",
       "      <td>rs17844444</td>\n",
       "      <td>PCDHB6</td>\n",
       "      <td>verapamil</td>\n",
       "      <td>33829662</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>\"There are 3 highly significant SNPs (p‐value ...</td>\n",
       "      <td>Allele A is associated with decreased clinical...</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11600</th>\n",
       "      <td>rs2230433</td>\n",
       "      <td>ITGAL</td>\n",
       "      <td>verapamil</td>\n",
       "      <td>33829662</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>\"There are 3 highly significant SNPs (p‐value ...</td>\n",
       "      <td>Allele C is associated with increased clinical...</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11601</th>\n",
       "      <td>rs3733694</td>\n",
       "      <td>PCDHB8</td>\n",
       "      <td>verapamil</td>\n",
       "      <td>33829662</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>\"There are 3 highly significant SNPs (p‐value ...</td>\n",
       "      <td>Allele G is associated with decreased clinical...</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11602 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     dbSNP_ID GENE_SYMBOL   DRUG_NAME  \\\n",
       "0                         CYP3A4*1, CYP3A4*17      CYP3A4  nifedipine   \n",
       "1                                   rs5031016      CYP2A6    warfarin   \n",
       "2      CYP2D6 ultrarapid metabolizer genotype      CYP2D6    tramadol   \n",
       "3                                   CYP2C9*11      CYP2C9    warfarin   \n",
       "4                         CYP2B6*1, CYP2B6*18      CYP2B6   efavirenz   \n",
       "...                                       ...         ...         ...   \n",
       "11597                               rs2242480      CYP3A4  tacrolimus   \n",
       "11598                               rs7590387       RAMP1    erenumab   \n",
       "11599                              rs17844444      PCDHB6   verapamil   \n",
       "11600                               rs2230433       ITGAL   verapamil   \n",
       "11601                               rs3733694      PCDHB8   verapamil   \n",
       "\n",
       "           PMID    PHENOTYPE_CATEGORY SIGNIFICANCE  \\\n",
       "0      15634941  Other, Metabolism/PK   not stated   \n",
       "1      22248286                Dosage           no   \n",
       "2      18204346         Metabolism/PK          yes   \n",
       "3      33350885                Dosage   not stated   \n",
       "4      16495778         Metabolism/PK          yes   \n",
       "...         ...                   ...          ...   \n",
       "11597  38004558         Metabolism/PK          yes   \n",
       "11598  34965002              Efficacy           no   \n",
       "11599  33829662              Efficacy          yes   \n",
       "11600  33829662              Efficacy          yes   \n",
       "11601  33829662              Efficacy          yes   \n",
       "\n",
       "                                                   NOTES  \\\n",
       "0      in vitro expression of the recombinant CYP3A4*...   \n",
       "1      No association was found between this variant ...   \n",
       "2      Median (+)R,R-tramadol area under the curve wa...   \n",
       "3      \"This case suggests that CYP2C9 *11/*11 carrie...   \n",
       "4      Please note that in the paper the allele was r...   \n",
       "...                                                  ...   \n",
       "11597  \"CYP3A4*18B genotypes were also significant co...   \n",
       "11598  \"RAMP1 rs7590387 was found to confer a lower p...   \n",
       "11599  \"There are 3 highly significant SNPs (p‐value ...   \n",
       "11600  \"There are 3 highly significant SNPs (p‐value ...   \n",
       "11601  \"There are 3 highly significant SNPs (p‐value ...   \n",
       "\n",
       "                                                SENTENCE ALLELE_PharmGKB  \n",
       "0      CYP3A4 *17 is associated with decreased metabo...             *17  \n",
       "1      Allele G is not associated with increased dose...               G  \n",
       "2      CYP2D6 ultra-metabolizer genotype is associate...             NaN  \n",
       "3      CYP2C9 *11/*11 is associated with decreased do...         *11/*11  \n",
       "4      CYP2B6 *1/*18 is associated with increased con...          *1/*18  \n",
       "...                                                  ...             ...  \n",
       "11597  Genotypes CT + TT is associated with increased...         CT + TT  \n",
       "11598  Allele G is associated with decreased clinical...               G  \n",
       "11599  Allele A is associated with decreased clinical...               A  \n",
       "11600  Allele C is associated with increased clinical...               C  \n",
       "11601  Allele G is associated with decreased clinical...               G  \n",
       "\n",
       "[11602 rows x 9 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# renaming relavant column names\n",
    "col_name_change = {'Variant/Haplotypes': 'dbSNP_ID',\n",
    "                  'Gene': 'GENE_SYMBOL',\n",
    "                  'Drug(s)': 'DRUG_NAME',\n",
    "                  'Phenotype Category': 'PHENOTYPE_CATEGORY',\n",
    "                  'Significance': 'SIGNIFICANCE',\n",
    "                  'Notes': 'NOTES',\n",
    "                  'Sentence': 'SENTENCE',\n",
    "                  'Alleles': 'ALLELE_PharmGKB'}\n",
    "\n",
    "drug_df.rename(columns=col_name_change, inplace=True)\n",
    "drug_df_subset = drug_df[['dbSNP_ID','GENE_SYMBOL', 'DRUG_NAME', 'PMID', 'PHENOTYPE_CATEGORY', 'SIGNIFICANCE', 'NOTES', 'SENTENCE', 'ALLELE_PharmGKB']]\n",
    "drug_df_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbSNP_ID</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>PMID</th>\n",
       "      <th>PHENOTYPE_CATEGORY</th>\n",
       "      <th>SIGNIFICANCE</th>\n",
       "      <th>NOTES</th>\n",
       "      <th>SENTENCE</th>\n",
       "      <th>ALLELE_PharmGKB</th>\n",
       "      <th>ALLELE_23andme</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs11807862</td>\n",
       "      <td>PRDM16</td>\n",
       "      <td>atorvastatin, hmg coa reductase inhibitors, si...</td>\n",
       "      <td>24096969</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>no</td>\n",
       "      <td>There could be strand confusion with this A/T ...</td>\n",
       "      <td>Genotype AA is associated with decreased respo...</td>\n",
       "      <td>AA</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs228729</td>\n",
       "      <td>PER3</td>\n",
       "      <td>lithium</td>\n",
       "      <td>21781277</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Allele T is not associated with increased resp...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs228642</td>\n",
       "      <td>PER3</td>\n",
       "      <td>lithium</td>\n",
       "      <td>21781277</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Allele C is not associated with increased resp...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs228666</td>\n",
       "      <td>PER3</td>\n",
       "      <td>lithium</td>\n",
       "      <td>21781277</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Allele C is not associated with increased resp...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs228697</td>\n",
       "      <td>PER3</td>\n",
       "      <td>lithium</td>\n",
       "      <td>21781277</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Allele G is not associated with increased resp...</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5664</th>\n",
       "      <td>rs3810651</td>\n",
       "      <td>GABRQ</td>\n",
       "      <td>botulinum toxin type a</td>\n",
       "      <td>31014225</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>no</td>\n",
       "      <td>No significant difference in allele frequency ...</td>\n",
       "      <td>Allele T is not associated with response to bo...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5665</th>\n",
       "      <td>rs17435</td>\n",
       "      <td>MECP2</td>\n",
       "      <td>cisplatin, fluorouracil, mitoxantrone</td>\n",
       "      <td>21635146</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>The study described this variant within the ME...</td>\n",
       "      <td>Allele A is associated with response to cispla...</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5666</th>\n",
       "      <td>rs1734787</td>\n",
       "      <td>MECP2</td>\n",
       "      <td>cisplatin, fluorouracil, mitoxantrone</td>\n",
       "      <td>21635146</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>The study described this variant within the ME...</td>\n",
       "      <td>Allele A is associated with response to cispla...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5667</th>\n",
       "      <td>rs1734791</td>\n",
       "      <td>MECP2</td>\n",
       "      <td>cisplatin, fluorouracil, mitoxantrone</td>\n",
       "      <td>21635146</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>The study described this variant within the ME...</td>\n",
       "      <td>Allele A is associated with response to cispla...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5668</th>\n",
       "      <td>rs1050828</td>\n",
       "      <td>G6PD</td>\n",
       "      <td>artemisinin and derivatives</td>\n",
       "      <td>22940027</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>no</td>\n",
       "      <td>G6PD deficient (those with genotype TT, TC, or...</td>\n",
       "      <td>Allele T is not associated with response to ar...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5669 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        dbSNP_ID GENE_SYMBOL  \\\n",
       "0     rs11807862      PRDM16   \n",
       "1       rs228729        PER3   \n",
       "2       rs228642        PER3   \n",
       "3       rs228666        PER3   \n",
       "4       rs228697        PER3   \n",
       "...          ...         ...   \n",
       "5664   rs3810651       GABRQ   \n",
       "5665     rs17435       MECP2   \n",
       "5666   rs1734787       MECP2   \n",
       "5667   rs1734791       MECP2   \n",
       "5668   rs1050828        G6PD   \n",
       "\n",
       "                                              DRUG_NAME      PMID  \\\n",
       "0     atorvastatin, hmg coa reductase inhibitors, si...  24096969   \n",
       "1                                               lithium  21781277   \n",
       "2                                               lithium  21781277   \n",
       "3                                               lithium  21781277   \n",
       "4                                               lithium  21781277   \n",
       "...                                                 ...       ...   \n",
       "5664                             botulinum toxin type a  31014225   \n",
       "5665              cisplatin, fluorouracil, mitoxantrone  21635146   \n",
       "5666              cisplatin, fluorouracil, mitoxantrone  21635146   \n",
       "5667              cisplatin, fluorouracil, mitoxantrone  21635146   \n",
       "5668                        artemisinin and derivatives  22940027   \n",
       "\n",
       "     PHENOTYPE_CATEGORY SIGNIFICANCE  \\\n",
       "0              Efficacy           no   \n",
       "1              Efficacy           no   \n",
       "2              Efficacy           no   \n",
       "3              Efficacy           no   \n",
       "4              Efficacy           no   \n",
       "...                 ...          ...   \n",
       "5664           Efficacy           no   \n",
       "5665           Efficacy          yes   \n",
       "5666           Efficacy          yes   \n",
       "5667           Efficacy          yes   \n",
       "5668           Efficacy           no   \n",
       "\n",
       "                                                  NOTES  \\\n",
       "0     There could be strand confusion with this A/T ...   \n",
       "1                                                   NaN   \n",
       "2                                                   NaN   \n",
       "3                                                   NaN   \n",
       "4                                                   NaN   \n",
       "...                                                 ...   \n",
       "5664  No significant difference in allele frequency ...   \n",
       "5665  The study described this variant within the ME...   \n",
       "5666  The study described this variant within the ME...   \n",
       "5667  The study described this variant within the ME...   \n",
       "5668  G6PD deficient (those with genotype TT, TC, or...   \n",
       "\n",
       "                                               SENTENCE ALLELE_PharmGKB  \\\n",
       "0     Genotype AA is associated with decreased respo...              AA   \n",
       "1     Allele T is not associated with increased resp...               T   \n",
       "2     Allele C is not associated with increased resp...               C   \n",
       "3     Allele C is not associated with increased resp...               C   \n",
       "4     Allele G is not associated with increased resp...               G   \n",
       "...                                                 ...             ...   \n",
       "5664  Allele T is not associated with response to bo...               T   \n",
       "5665  Allele A is associated with response to cispla...               A   \n",
       "5666  Allele A is associated with response to cispla...               A   \n",
       "5667  Allele A is associated with response to cispla...               A   \n",
       "5668  Allele T is not associated with response to ar...               T   \n",
       "\n",
       "     ALLELE_23andme  \n",
       "0                 T  \n",
       "1                 T  \n",
       "2                 C  \n",
       "3                 T  \n",
       "4                 C  \n",
       "...             ...  \n",
       "5664              T  \n",
       "5665              T  \n",
       "5666              A  \n",
       "5667              A  \n",
       "5668              C  \n",
       "\n",
       "[5669 rows x 10 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merging and only taking necessary columns \n",
    "merged_df = pd.merge(twentythree_df, drug_df_subset, on='dbSNP_ID')\n",
    "result_df = merged_df[['dbSNP_ID', 'GENE_SYMBOL', 'DRUG_NAME', 'PMID', 'PHENOTYPE_CATEGORY', 'SIGNIFICANCE', 'NOTES', 'SENTENCE', 'ALLELE_PharmGKB', 'ALLELE_23andme']]\n",
    "\n",
    "result_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Q2. Filter the output so that it only contains significant associations (SIGNIFICANCE is yes) for variants that affect the drug efficacy (`PHENOTYPE_CATEGORY` is `efficacy`).\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbSNP_ID</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>PMID</th>\n",
       "      <th>PHENOTYPE_CATEGORY</th>\n",
       "      <th>SIGNIFICANCE</th>\n",
       "      <th>NOTES</th>\n",
       "      <th>SENTENCE</th>\n",
       "      <th>ALLELE_PharmGKB</th>\n",
       "      <th>ALLELE_23andme</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rs1801131</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>oxaliplatin</td>\n",
       "      <td>20385995</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>Statistics given as trend for increased overal...</td>\n",
       "      <td>Genotypes GG + GT are associated with increase...</td>\n",
       "      <td>GG + GT</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rs1801131</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Vitamin B-complex, Incl. Combinations</td>\n",
       "      <td>27035272</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>Please note that allele has been complemented ...</td>\n",
       "      <td>Allele G is associated with increased response...</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>rs1801131</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>methotrexate</td>\n",
       "      <td>25618758</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>Alleles given as C and A. Efficacy of treatmen...</td>\n",
       "      <td>Genotype GT is associated with decreased respo...</td>\n",
       "      <td>GT</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>rs1801131</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>methotrexate</td>\n",
       "      <td>23198157</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>Patients with the GG genotype had better overa...</td>\n",
       "      <td>Genotype GG is associated with increased respo...</td>\n",
       "      <td>GG</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>rs1801131</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>methotrexate</td>\n",
       "      <td>24386571</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>Patients with the TT genotype had significantl...</td>\n",
       "      <td>Genotype TT is associated with decreased respo...</td>\n",
       "      <td>TT</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5660</th>\n",
       "      <td>rs502434</td>\n",
       "      <td>GRIA3</td>\n",
       "      <td>Selective serotonin reuptake inhibitors, venla...</td>\n",
       "      <td>23394390</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>Regression model looking at interaction betwee...</td>\n",
       "      <td>Genotype CC is associated with decreased respo...</td>\n",
       "      <td>CC</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5663</th>\n",
       "      <td>rs3810651</td>\n",
       "      <td>GABRQ</td>\n",
       "      <td>venlafaxine</td>\n",
       "      <td>23394390</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>Those with the AA/AT genotype are more likely ...</td>\n",
       "      <td>Genotypes AA + AT are associated with increase...</td>\n",
       "      <td>AA + AT</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5665</th>\n",
       "      <td>rs17435</td>\n",
       "      <td>MECP2</td>\n",
       "      <td>cisplatin, fluorouracil, mitoxantrone</td>\n",
       "      <td>21635146</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>The study described this variant within the ME...</td>\n",
       "      <td>Allele A is associated with response to cispla...</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5666</th>\n",
       "      <td>rs1734787</td>\n",
       "      <td>MECP2</td>\n",
       "      <td>cisplatin, fluorouracil, mitoxantrone</td>\n",
       "      <td>21635146</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>The study described this variant within the ME...</td>\n",
       "      <td>Allele A is associated with response to cispla...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5667</th>\n",
       "      <td>rs1734791</td>\n",
       "      <td>MECP2</td>\n",
       "      <td>cisplatin, fluorouracil, mitoxantrone</td>\n",
       "      <td>21635146</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>The study described this variant within the ME...</td>\n",
       "      <td>Allele A is associated with response to cispla...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1311 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       dbSNP_ID GENE_SYMBOL  \\\n",
       "8     rs1801131       MTHFR   \n",
       "9     rs1801131       MTHFR   \n",
       "11    rs1801131       MTHFR   \n",
       "14    rs1801131       MTHFR   \n",
       "15    rs1801131       MTHFR   \n",
       "...         ...         ...   \n",
       "5660   rs502434       GRIA3   \n",
       "5663  rs3810651       GABRQ   \n",
       "5665    rs17435       MECP2   \n",
       "5666  rs1734787       MECP2   \n",
       "5667  rs1734791       MECP2   \n",
       "\n",
       "                                              DRUG_NAME      PMID  \\\n",
       "8                                           oxaliplatin  20385995   \n",
       "9                 Vitamin B-complex, Incl. Combinations  27035272   \n",
       "11                                         methotrexate  25618758   \n",
       "14                                         methotrexate  23198157   \n",
       "15                                         methotrexate  24386571   \n",
       "...                                                 ...       ...   \n",
       "5660  Selective serotonin reuptake inhibitors, venla...  23394390   \n",
       "5663                                        venlafaxine  23394390   \n",
       "5665              cisplatin, fluorouracil, mitoxantrone  21635146   \n",
       "5666              cisplatin, fluorouracil, mitoxantrone  21635146   \n",
       "5667              cisplatin, fluorouracil, mitoxantrone  21635146   \n",
       "\n",
       "     PHENOTYPE_CATEGORY SIGNIFICANCE  \\\n",
       "8              Efficacy          yes   \n",
       "9              Efficacy          yes   \n",
       "11             Efficacy          yes   \n",
       "14             Efficacy          yes   \n",
       "15             Efficacy          yes   \n",
       "...                 ...          ...   \n",
       "5660           Efficacy          yes   \n",
       "5663           Efficacy          yes   \n",
       "5665           Efficacy          yes   \n",
       "5666           Efficacy          yes   \n",
       "5667           Efficacy          yes   \n",
       "\n",
       "                                                  NOTES  \\\n",
       "8     Statistics given as trend for increased overal...   \n",
       "9     Please note that allele has been complemented ...   \n",
       "11    Alleles given as C and A. Efficacy of treatmen...   \n",
       "14    Patients with the GG genotype had better overa...   \n",
       "15    Patients with the TT genotype had significantl...   \n",
       "...                                                 ...   \n",
       "5660  Regression model looking at interaction betwee...   \n",
       "5663  Those with the AA/AT genotype are more likely ...   \n",
       "5665  The study described this variant within the ME...   \n",
       "5666  The study described this variant within the ME...   \n",
       "5667  The study described this variant within the ME...   \n",
       "\n",
       "                                               SENTENCE ALLELE_PharmGKB  \\\n",
       "8     Genotypes GG + GT are associated with increase...         GG + GT   \n",
       "9     Allele G is associated with increased response...               G   \n",
       "11    Genotype GT is associated with decreased respo...              GT   \n",
       "14    Genotype GG is associated with increased respo...              GG   \n",
       "15    Genotype TT is associated with decreased respo...              TT   \n",
       "...                                                 ...             ...   \n",
       "5660  Genotype CC is associated with decreased respo...              CC   \n",
       "5663  Genotypes AA + AT are associated with increase...         AA + AT   \n",
       "5665  Allele A is associated with response to cispla...               A   \n",
       "5666  Allele A is associated with response to cispla...               A   \n",
       "5667  Allele A is associated with response to cispla...               A   \n",
       "\n",
       "     ALLELE_23andme  \n",
       "8                 T  \n",
       "9                 T  \n",
       "11                T  \n",
       "14                T  \n",
       "15                T  \n",
       "...             ...  \n",
       "5660              T  \n",
       "5663              T  \n",
       "5665              T  \n",
       "5666              A  \n",
       "5667              A  \n",
       "\n",
       "[1311 rows x 10 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_df = result_df[(result_df['SIGNIFICANCE'] == 'yes') & (result_df['PHENOTYPE_CATEGORY'] == 'Efficacy')]\n",
    "filtered_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Q3. Save the output of the filtering step in a tab-separated file (23andme_PharmGKB_map.tsv) with the following columns:\n",
    "dbSNP_ID, GENE_SYMBOL, DRUG_NAME, NOTES, SENTENCE, ALLELE_PharmGKB, ALLELE_23andme"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_subset = filtered_df[['dbSNP_ID', 'GENE_SYMBOL', 'DRUG_NAME', 'NOTES', 'SENTENCE', 'ALLELE_PharmGKB', 'ALLELE_23andme']]\n",
    "filtered_subset.to_csv('23andme_PharmGKB_map.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Q4. Create a tab separated file (23andme_PharmGKB_summary.tsv) with summarized data with the following columns:\n",
    "GENE_SYMBOL, DRUG_NAME, dbSNP_IDs (list of IDs separated by \";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_tsv = filtered_df[['GENE_SYMBOL', 'DRUG_NAME', 'dbSNP_ID']]\n",
    "summary_tsv.to_csv('23andme_PharmGKB_summary.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Q5. Plot the distribution (histogram) of the number of drugs associated with a gene, and the number of SNPs for a gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>dbSNP_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCA1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABCC1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABCC2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCC4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>410</th>\n",
       "      <td>XRCC3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411</th>\n",
       "      <td>XRCC4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412</th>\n",
       "      <td>ZNF432</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>413</th>\n",
       "      <td>ZNF697</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>414</th>\n",
       "      <td>ZNF804A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>415 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    GENE_SYMBOL  dbSNP_ID\n",
       "0         ABCA1         2\n",
       "1         ABCB1         4\n",
       "2         ABCC1         6\n",
       "3         ABCC2         2\n",
       "4         ABCC4         2\n",
       "..          ...       ...\n",
       "410       XRCC3         1\n",
       "411       XRCC4         1\n",
       "412      ZNF432         1\n",
       "413      ZNF697         1\n",
       "414     ZNF804A         1\n",
       "\n",
       "[415 rows x 2 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# number of snps associated with each gene\n",
    "gene_snp_counts = filtered_subset.groupby('GENE_SYMBOL')['dbSNP_ID'].nunique().reset_index()\n",
    "gene_snp_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCA1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABCC1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABCC2</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCC4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>410</th>\n",
       "      <td>XRCC3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411</th>\n",
       "      <td>XRCC4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412</th>\n",
       "      <td>ZNF432</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>413</th>\n",
       "      <td>ZNF697</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>414</th>\n",
       "      <td>ZNF804A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>415 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    GENE_SYMBOL  DRUG_NAME\n",
       "0         ABCA1          2\n",
       "1         ABCB1         16\n",
       "2         ABCC1          2\n",
       "3         ABCC2          8\n",
       "4         ABCC4          2\n",
       "..          ...        ...\n",
       "410       XRCC3          2\n",
       "411       XRCC4          1\n",
       "412      ZNF432          1\n",
       "413      ZNF697          1\n",
       "414     ZNF804A          1\n",
       "\n",
       "[415 rows x 2 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# number of drugs associated with each gene\n",
    "gene_drug_counts = filtered_subset.groupby('GENE_SYMBOL')['DRUG_NAME'].nunique().reset_index()\n",
    "gene_drug_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
